provision: implement reservation expiration

zaibon opened the pull request:
the provision modules now keeps track of the reservation currently
deployed on the node.
It then check periodically which reservation have expired and
decomission them from the system.

fixes #121

zaibon committed:
provision: implement reservation expiration

the provision modules now keeps track of the reservation currently
deployed on the node.
It then check periodically which reservation have expired and
decomission them from the system.

fixes #121

zaibon commented:
I open this as a draft cause there are couple of things I'm too happy about.
While this PR is fully working I would like to discuss some point:

- In this PR I now keep the full reservation object in the local_store of provisiond. I do that cause to decommission networks I need to have the full network object
- the local store has only a memory backend at the moment, we need to move that on disk
- I didn't handle cases when provisiond process restarts
- The concurrent logic for the provision engine is a bit rough, I think we can improve that part
- The value for expiration check is hard-coded and arbitrary
- Most of the new code have no tests

One thing I notice while writing this is that both `iterExpiration` and `iterReservation` function have the same signature `func ctx context.Context) <-chan Reservation` which is the same as the only method in the `ReservationSource` interface
```go
type ReservationSource interface {
	Reservations(ctx context.Context) <-chan Reservation
}
```

so actually we could maybe improve the `combinedResource` struct:
```go
type compinedSource struct {
	s1 ReservationSource
	s2 ReservationSource
}
```
to support any number of source by implementing a proper fade-in and then just make `iterExpiration` and `iterReservation` a source for the engine.

zaibon committed:
travis: do run the linting tools

zaibon committed:
go mod tidy

muhamadazmy commented on modules/provision/container.go:
```diff
@@ -150,3 +150,10 @@ func ContainerProvision(ctx context.Context, reservation Reservation) (interface
 	return id, nil
 }
+
+func containerDecomission(ctx context.Context, reservation Reservation) error {
```
`Decommission` not `Decomission`

muhamadazmy commented on modules/provision/container.go:
```diff
@@ -150,3 +150,10 @@ func ContainerProvision(ctx context.Context, reservation Reservation) (interface
+	client := GetZBus(ctx)
+
+	containerClient := stubs.NewContainerModuleStub(client)
+	return containerClient.Delete(reservation.User, modules.ContainerID(reservation.ID))
```
U also need to unmount the rootfs of the container, so I guess u will need to do inspect first, store the rootfs, delete the container then unmount the filesystem.

muhamadazmy commented on modules/provision/engine.go:
```diff
@@ -18,40 +20,141 @@ type defaultEngine struct {
+
+		case reservation, ok := <-cReservation:
+			if !ok {
+				log.Info().Msg("reservation source is emptied. stopping engine")
```
Why do we need this. I assume a reservation source should never considered "terminated", it can just have no new stuff for sometime.

zaibon replied:
If we don't handle that case and for any reason the source provider close its channel the loop on the select will be an infinite loop receiving nil reservations. Which will eventually crash the daemon.
I had the case when working on this. When provisiond boots, it first provide all the reservations for this node once, then starts waiting for new one. I now changed the logic so I don't use a different source for this, but the underlying problem remains.

muhamadazmy commented on modules/provision/engine.go:
```diff
@@ -18,40 +20,141 @@ type defaultEngine struct {
-	return nil
+}
+
+func (e *defaultEngine) iterReservation(ctx context.Context) <-chan Reservation {
```
I don't get why u need this method. I can see that you only copy the objects from the source channel to another channel, why not use the e.source.Reservations(ctx) channel directly ?

> If u need to do call to the `validate()` i think this can be done easier in the source itself.

zaibon replied:
yep I can move the validation in the source itself, and get rid of this function.

muhamadazmy commented on modules/provision/engine.go:
```diff
@@ -18,40 +20,141 @@ type defaultEngine struct {
+	return c
+}
+
+func (e *defaultEngine) iterExpiration(ctx context.Context) <-chan *Reservation {
```
I think here u can also create a `ReservationSource` to wrap this method. It will be cleaner to have this logic in a different source than directly on the engine. Because that's a lot of logic that doesn't directly relate to the engine.

zaibon committed:
provision: split big interfaces into smaller one

In this commit I mostly re-organized how we get and store reservations
Do do this I splitted the big interfaces we had local RemoteStore and
LocaleStore into mostly one method interfaces that I can then more
easily re-use for different use case

The LRU cache is now gone since we do keep a copy of the reservation
locally anyhow. The owner cache just use the local copy of the
reservation.

The reservations sources have also been cleaned up and the goroutine
that sends reservation to be decommisioned also now implement the
ReservationSource interface, so the engine really have only one source
for any type of reservation

muhamadazmy commented on modules/provision/engine.go:
```diff
@@ -17,41 +23,87 @@ type defaultEngine struct {
+			log.Info().Msg("provision engine context done, exiting")
+			return nil
+
+		case reservation, ok := <-cReservation:
```
Just some cosmetics here, this switch case is so long, can we move the logic to a separate function(s) instead of one big select case.

muhamadazmy commented on modules/provision/local_store.go:
```diff
@@ -0,0 +1,81 @@
+	return output, nil
+}
+
+func isExpired(r *Reservation) bool {
```
Why this is not defined as a member method on the Reservation itself? 
like
```suggestion
func (r *Reservation) Expired() bool {
```

zaibon committed:
provision: unify usage of pointer for the reservation object

we now use pointer everywhere, cause the reservation for a network can
be a huge object, we don't want to copy this over

zaibon committed:
provision: split engine logic into 2 methods

zaibon committed:
fix linting errors and go mod tidy

zaibon committed:
more linting error fixes
